fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java

```diff
@@ -235,6 +240,110 @@ public Transaction ensureActive() {
         return transaction;
     }
 
+    /**
+     * Set the read version used by this transaction. All reads to the database will include
+     * only changes that were committed at this version or smaller. A transaction's read version
+     * can only be set once, so if this function is called multiple times, it will return the
+     * previously set read version. If this method is called and another caller has already called
+     * {@link #getReadVersionAsync()} or {@link #getReadVersion()}, this method may throw a
+     * {@link RecordCoreException} indicating that there is already an outstanding read version request
+     * if that request has not yet completed.
+     *
+     * @param readVersion the read version this transaction should use if is not already set
+     * @return this transaction's read version
+     * @see Transaction#setReadVersion(long)
+     */
+    public synchronized long setReadVersion(long readVersion) {
+        if (hasReadVersion()) {
+            return this.readVersion;
+        }
+        if (readVersionFuture != null) {
+            if (MoreAsyncUtil.isCompletedNormally(readVersionFuture)) {
+                return joinNow(readVersionFuture);
+            } else {
+                throw new RecordCoreException("Cannot set read version as read version request is outstanding");
+            }
+        }
+        ensureActive().setReadVersion(readVersion);
+        this.readVersion = readVersion;
+        this.readVersionFuture = CompletableFuture.completedFuture(readVersion);
+        return readVersion;
+    }
+
+    /**
+     * Get the read version used by this transaction. All reads to the database will include only changes that
+     * were committed at this version or smaller. If the read version has not already been set, this
+     * may require talking to the database. If the read version has already been set, then this will return
+     * with an already completed future.
+     *
+     * <p>
+     * Note that this method is {@code synchronized}, but only creating the future (<em>not</em> waiting on
+     * the future) will block other threads. Thus, while it is advised that this method ony be called once
+     * and by only one caller at a time, if it safe to use this method in asynchronous contexts.
+     * </p>
+     *
+     * @return a future that will contain the read version of this transaction
+     * @see Transaction#getReadVersion()
+     */
+    @Nonnull
+    public synchronized CompletableFuture<Long> getReadVersionAsync() {
+        if (readVersionFuture != null) {
+            return readVersionFuture;
+        }
+        final Transaction tr = ensureActive();
```

Comment:
I added a test case for this. The exact behavior is that doing something like that will do one of the following things:

1. There is a successfully returned read version, if the race is run exactly the right way.
1. An `FDBException` is thrown from the FDB C library saying that the operation was cancelled.
1. An `IllegalStateException` is thrown from the FDB Java bindings saying that the transaction is already closed (essentially).
1. One gets a `RecordCoreStoreageException("Transaction is no longer active.")` exception (not in my test, but in general if one is concurrently calling "close" an `getReadVersionAsync`.

I'm not sure exactly what the right thing is. I've moved it so that there is an `ensureActive` call right before the `getReadVersion` call, but I'm not totally sure it's a big problem. I think other operations probably have the same race condition (in that if you call `close` after they've already gotten their `transaction`, you'll see a different error message), and there is still a gap here. I'm inclined to think we should consider them all equivalent enough, but maybe I'm wrong.

Reply:
As a side note, calling `ensureActive` in the lambda I think means it doesn't need to have its own closure (necessarily), so JVMs may decide to optimize it better, so it is probably worth the change anyway.